drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2494 @@
+
+static int nvs_close_ate_valid(FAR struct nvs_fs *fs,
+                               FAR const struct nvs_ate *entry)
+{
+  if ((!nvs_ate_valid(fs, entry)) || (entry->len != 0) ||
+      (entry->id != NVS_SPECIAL_ATE_ID))
+    {
+      return 0;
+    }
+
+  if ((fs->geo.erasesize - entry->offset) % sizeof(struct nvs_ate))
+    {
+      return 0;
+    }
+
+  return 1;
+}
+
+/****************************************************************************
+ * Name: nvs_flash_wrt_entry
+ *
+ * Description:
+ *   store an entry in flash
+ *
+ ****************************************************************************/
+
+static int nvs_flash_wrt_entry(FAR struct nvs_fs *fs, uint32_t id,
+            FAR const uint8_t *key, size_t key_size, FAR const void *data,
+            size_t len)
+{
+  int rc;
+  struct nvs_ate entry;
+
+  memset(&entry, CONFIG_MTD_CONFIG_ERASEDVALUE, sizeof(entry));
+  entry.id = id;
+  entry.offset = (uint16_t)(fs->data_wra & ADDR_OFFS_MASK);
+  entry.len = (uint16_t)len;
+  entry.key_len = (uint16_t)key_size;
+
+  nvs_ate_crc8_update(&entry);
+
+  /* let's sew key and data into one, key comes first, then data */
+
+  rc = nvs_flash_data_wrt(fs, key, key_size);
+  if (rc)
+    {
+      ferr("Write key failed, rc=%d\n", rc);
+      return rc;
+    }
+
+  rc = nvs_flash_data_wrt(fs, data, len);
+  if (rc)
+    {
+      ferr("Write value failed, rc=%d\n", rc);
+      return rc;
+    }
+
+  rc = nvs_flash_ate_wrt(fs, &entry);
+  if (rc)
+    {
+      ferr("Write ate failed, rc=%d\n", rc);
+      return rc;
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: nvs_recover_last_ate
+ *
+ * Description:
+ *   If the closing ate is invalid, its offset cannot be trusted and
+ *   the last valid ate of the block should instead try to be recovered
+ *   by going through all ate's.
+ *
+ *   addr should point to the faulty closing ate and will be updated to
+ *   the last valid ate. If no valid ate is found it will be left untouched.
+ *
+ ****************************************************************************/
+
+static int nvs_recover_last_ate(FAR struct nvs_fs *fs,
+                                FAR uint32_t *addr)
+{
+  uint32_t data_end_addr;
+  uint32_t ate_end_addr;
+  struct nvs_ate end_ate;
+  int rc;
+
+  finfo("Recovering last ate from block %" PRIu32 "\n",
+        (*addr >> ADDR_BLOCK_SHIFT));
+
+  *addr -= sizeof(struct nvs_ate);
+  ate_end_addr = *addr;
+  data_end_addr = *addr & ADDR_BLOCK_MASK;
+  while (ate_end_addr > data_end_addr)
+    {
+      rc = nvs_flash_ate_rd(fs, ate_end_addr, &end_ate);
+      if (rc)
+        {
+          return rc;
+        }
+
+      if (nvs_ate_valid(fs, &end_ate))
+        {
+          /* Found a valid ate, update data_end_addr and *addr */
+
+          data_end_addr &= ADDR_BLOCK_MASK;
+          data_end_addr += end_ate.offset + end_ate.key_len + end_ate.len;
+          *addr = ate_end_addr;
+        }
```

Comment:
We need to traverse all the ate, in case an invalid date is in the middle (Possible but unlikely).
For robustness, I think it is better to traverse all.